### DISCUSSION TITLE
Adding icons and making it executable

### DISCUSSION THREAD

**Discussion by rafinskipg** [nodejs/single-executable#67] [Q&A]
Hello. 

We are interested in the new SEA feature from node, but we want/need our apps to have some icon and be able to be executed with a user without the need of using the console.

Does anyone have any guidance on how we can achieve that? Any kind of idea / suggestion is appreciated, thanks

**Comment by RaisinTen** [comment]
I think it is possible to add an icon to an executable on Windows - https://stackoverflow.com/questions/673203/add-icon-to-existing-exe-file-from-the-command-line but I don't think it's possible on macOS or Linux. Do you know of any existing executables that have such custom icons?

> be able to be executed with a user without the need of using the console.

Could you describe your application a bit? From what I can tell, if you want to create an application that doesn't need to be executed from the console, it's probably a desktop app you're thinking about, in which case, you might be interested in building an app using electron (or some other tool for building desktop apps).

**Reply by ljharb** [reply]
You can add an icon to an executable on Macs by getting info on it, clicking on the icon, and pasting.

**Reply by RaisinTen** [reply]
@ljharb in case you're referring to https://superuser.com/a/37813, I don't think the new icon will persist if you distribute the executable. In fact, you can even try to upload the file somewhere and download it on the same macOS system you uploaded it from and you'll see that it would have the same icon it had before the custom icon was added.

**Reply by rafinskipg** [reply]
hello @RaisinTen 

The use case is to allow users to launch a node-static simple http server that serves a webapp on their machine, (localhost:8080 for example). 

We can not use electro because it limits the user access to browser extensions, and we want the user to be able to access the locally run website with all their extensions.

We are thinking about distributed web applications that are self-hosted in your own machine), simple http server serving just a plain website. 

Ideally if the user could just download and install something that launches the webserver automatically, with a nice icon, that would be amazing.

**Reply by RaisinTen** [reply]
> The use case is to allow users to launch a node-static simple http server that serves a webapp on their machine, (localhost:8080 for example).

What's the UI like for closing the server when the user is done using it? Do they have to terminate the process (for example, from the task manager on Windows)? It feels like it would be a better UI if your application had a nicer way for users to terminate the server, which would typically be present in a desktop app, instead of an SEA.

> We can not use electro because it limits the user access to browser extensions, and we want the user to be able to access the locally run website with all their extensions.
> 
> We are thinking about distributed web applications that are self-hosted in your own machine), simple http server serving just a plain website.

What if your application opened the link in a browser, using [`shell.openExternal()`](https://www.electronjs.org/docs/latest/api/shell#shellopenexternalurl-options)? Since it's gonna run the browser that has the extensions, users will be able to use the extensions.

> Ideally if the user could just download and install something that launches the webserver automatically, with a nice icon, that would be amazing.

Unfortunately, I think the limitation is there at the OS-level because you as a developer can't add an icon on a single executable binary that would persist when you deploy it on all OSs. On macOS at least, you would need an application bundle (so it's a full-blown desktop app, and not an SEA anymore), i.e., a `<ApplicationName>.app` directory which would have an `Info.plist` file with the `CFBundleIconFile` key set to the path to the `<name>.icns` file, which is your application icon. There's a similar way for Linux, where you would need a desktop file, see - https://stackoverflow.com/questions/2126545/embedding-an-icon-in-a-linux-executable. Windows seems like the only one that allows you to actually embed the icon into your application.

So I think this falls out of the scope of SEAs in general.

**Reply by rafinskipg** [reply]
thanks for the context and the detailed explanation, yeah that makes a lot of sense @RaisinTen

**Reply by TessavWalstijn** [reply]
@RaisinTen an other use case is for simple tooling.

For example I have some tools currently build with the package pkg
They mostly use them and I like to have them branded to me or them with a custom icon.

Currently I haven't been able to use [`rcedit`](https://github.com/electron/rcedit) or [`Resource Hacker`](http://angusj.com/resourcehacker/) since those corrupts the executable.

NOTE: I haven't been able to try those tools out on the Node v20 generated .exe file.

However I would like to be able to include it within a "build pipeline" to include the icons instead of switching to an other program to polish the software.

**Reply by RaisinTen** [reply]
@TessavWalstijn I think long-term, most users will be building SEAs through tools that are composed of the building blocks we're developing like Node.js, postject, etc., so from a user's POV, it should feel like using one unified program.

**Reply by TessavWalstijn** [reply]
Makes sense to leaving that functionality to packages then.

**Reply by JoCat** [reply]
> Currently I haven't been able to use [`rcedit`](https://github.com/electron/rcedit) or [`Resource Hacker`](http://angusj.com/resourcehacker/) since those corrupts the executable.

Also encountered this problem, but with [resedit-cli](https://www.npmjs.com/package/resedit-cli)/[resedit](https://www.npmjs.com/package/resedit) everything works as it should
Example of use (resedit-cli): `resedit -i example.exe -o example.exe --icon 1,resources/icon.ico`

If you need to hide the console window, you can [try this](https://github.com/AuroraTeam/Glacier/blob/8b19e46af7ab60c7df7e367955cd7b556cfb2645/resedit.mjs#L10-L11).

**Reply by crisdosaygo** [reply]
@JoCat and @rafinskipg just tried resedit-cli - it works! The person who created it is ia genius. It's fantastic.
